homeassistant/components/camera/proxy.py

```diff
@@ -0,0 +1,264 @@
+    vol.Optional(CONF_CACHE_IMAGES, False): cv.boolean,
+    vol.Optional(CONF_MAX_STREAM_WIDTH): int,
+    vol.Optional(CONF_STREAM_QUALITY): int,
+})
+
+
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Set up the Proxy camera platform."""
+    async_add_devices([ProxyCamera(hass, config)])
+
+    return True
+
+
+@asyncio.coroutine
+def _read_frame(req):
+    """Read a single frame from an MJPEG stream."""
+    # based on https://gist.github.com/russss/1143799
+    import cgi
+    # Read in HTTP headers:
+    stream = req.content
+    # multipart/x-mixed-replace; boundary=--frameboundary
+    _mimetype, options = cgi.parse_header(req.headers['content-type'])
+    boundary = options.get('boundary').encode('utf-8')
+    if not boundary:
+        _LOGGER.error("Malformed MJPEG missing boundary")
+        raise Exception("Can't find content-type")
+
+    line = yield from stream.readline()
+    # Seek ahead to the first chunk
+    while line.strip() != boundary:
+        line = yield from stream.readline()
+    # Read in chunk headers
+    while line.strip() != b'':
+        parts = line.split(b':')
+        if len(parts) > 1 and parts[0].lower() == b'content-length':
+            # Grab chunk length
+            length = int(parts[1].strip())
+        line = yield from stream.readline()
+    image = yield from stream.read(length)
+    return image
+
+
+def _resize_image(image, opts):
+    """Resize image."""
+    from PIL import Image
+    import io
+
+    if not opts:
+        return image
+
+    quality = opts.quality if opts.quality else DEFAULT_QUALITY
+    new_width = opts.max_width
+
+    img = Image.open(io.BytesIO(image))
+    imgfmt = str(img.format)
+    if imgfmt != 'PNG' and imgfmt != 'JPEG':
+        _LOGGER.debug("Image is of unsupported type: %s", imgfmt)
+        return image
+
+    (old_width, old_height) = img.size
+    old_size = len(image)
+    if old_width <= new_width:
+        if opts.quality is None:
+            _LOGGER.debug("Image is smaller-than / equal-to requested width")
+            return image
+        new_width = old_width
+
+    scale = new_width / float(old_width)
+    new_height = int((float(old_height)*float(scale)))
+
+    img = img.resize((new_width, new_height), Image.ANTIALIAS)
+    imgbuf = io.BytesIO()
+    img.save(imgbuf, "JPEG", optimize=True, quality=quality)
+    newimage = imgbuf.getvalue()
+    if not opts.force_resize and len(newimage) >= old_size:
+        _LOGGER.debug("Using original image(%d bytes) "
+                      "because resized image (%d bytes) is not smaller",
+                      old_size, len(newimage))
+        return image
+
+    _LOGGER.debug("Resized image "
+                  "from (%dx%d - %d bytes) "
+                  "to (%dx%d - %d bytes)",
+                  old_width, old_height, old_size,
+                  new_width, new_height, len(newimage))
+    return newimage
+
+
+class ImageOpts():
```

Comment:
You can optionally use the `attrs` package which is installed as part of Home Assistant core

Reply:
Unless you are going to mandate it, I'd rather not.  As far as I can tell, it won't reduce any code for my use case, and would make the code harder to read.